relations/src/shielder/deposit.rs

```diff
@@ -1,149 +1,96 @@
+            types::{
+                BackendNullifier, BackendTokenAmount, BackendTokenId, BackendTrapdoor,
+                FrontendNullifier, FrontendTokenAmount, FrontendTokenId, FrontendTrapdoor,
+            },
+        },
+    };
+
+    #[circuit_field]
+    pub type CircuitField = crate::CircuitField;
+
+    #[relation_object_definition]
+    struct DepositRelation {
+        #[public_input(frontend_type = "FrontendNote", parse_with = "convert_note")]
+        pub note: BackendNote,
+        #[public_input(frontend_type = "FrontendTokenId")]
+        pub token_id: BackendTokenId,
+        #[public_input(frontend_type = "FrontendTokenAmount")]
+        pub token_amount: BackendTokenAmount,
+
+        // Private inputs
+        #[private_input(frontend_type = "FrontendTrapdoor")]
+        pub trapdoor: BackendTrapdoor,
+        #[private_input(frontend_type = "FrontendNullifier")]
+        pub nullifier: BackendNullifier,
     }
-}
 
-impl DepositRelation<FullInput> {
-    pub fn with_full_input(
-        note: FrontendNote,
-        token_id: FrontendTokenId,
-        token_amount: FrontendTokenAmount,
-        trapdoor: FrontendTrapdoor,
-        nullifier: FrontendNullifier,
-    ) -> Self {
-        DepositRelation {
-            note: Some(BackendNote::from(BigInteger256::new(note))),
-            token_id: Some(BackendTokenId::from(token_id)),
-            token_amount: Some(BackendTokenAmount::from(token_amount)),
-            trapdoor: Some(BackendTrapdoor::from(trapdoor)),
-            nullifier: Some(BackendNullifier::from(nullifier)),
-            _phantom: PhantomData,
-        }
-    }
-}
+    #[circuit_definition]
+    fn generate_constraints() {
+        let note = FpVar::new_input(ns!(cs, "note"), || self.note())?;
+        let token_id = FpVar::new_input(ns!(cs, "token id"), || self.token_id())?;
+        let token_amount = FpVar::new_input(ns!(cs, "token amount"), || self.token_amount())?;
 
-impl<S: State> ConstraintSynthesizer<CircuitField> for DepositRelation<S> {
-    fn generate_constraints(
-        self,
-        cs: ConstraintSystemRef<CircuitField>,
-    ) -> Result<(), SynthesisError> {
-        let note = FpVar::new_input(ns!(cs, "note"), || self.note.ok_or(AssignmentMissing))?;
-        let token_id = FpVar::new_input(ns!(cs, "token id"), || {
-            self.token_id.ok_or(AssignmentMissing)
-        })?;
-        let token_amount = FpVar::new_input(ns!(cs, "token amount"), || {
-            self.token_amount.ok_or(AssignmentMissing)
-        })?;
-
-        let trapdoor = FpVar::new_witness(ns!(cs, "trapdoor"), || {
-            self.trapdoor.ok_or(AssignmentMissing)
-        })?;
-        let nullifier = FpVar::new_witness(ns!(cs, "nullifier"), || {
-            self.nullifier.ok_or(AssignmentMissing)
-        })?;
+        let trapdoor = FpVar::new_witness(ns!(cs, "trapdoor"), || self.trapdoor())?;
+        let nullifier = FpVar::new_witness(ns!(cs, "nullifier"), || self.nullifier())?;
```

Comment:
I can see yet another chance for improvement here: we know which fields are public (and private) thanks to the macro - we can generate the "setup" (all the `FpVar::new_input` and `FpVar::new_witness` calls) in the macro itself and "just" require the developer to implement the validity checks.

Reply:
this would require some consideration how to handle it comfortably - sometimes, you have e.g. a vector of `FpVar`; also, the order of declaration is important

anyway - yes, you're right and we can think about it soon